File: 49-group-anagrams/49-group-anagrams.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        Map<String, List> map = new HashMap<>();
        for (int i = 0; i < strs.length; i++) {
            char[] ch = strs[i].toCharArray();
            Arrays.sort(ch);
            String key = String.valueOf(ch);
            System.out.println(key);
            if (!map.containsKey(key)) map.put(key, new ArrayList());
            map.get(key).add(strs[i]);
        }
        return new ArrayList(map.values());
    }
}

File: 49-group-anagrams/49-group-anagrams.scala
import scala.annotation.tailrec

object Solution {
    def groupAnagrams(strs: Array[String]): List[List[String]] = {
        @tailrec
    def go(i: Int, mp: Map[String, List[String]]): List[List[String]] = {
      if (i == strs.length) mp.values.toList
      else {
        val key = strs(i).sorted.mkString
        go(i + 1, mp + (key -> mp.get(key).map { value =>
          value :+ strs(i)
        }.getOrElse(List(strs(i)))))
      }
    }
    go(0, Map.empty[String, List[String]])
    }
}